File: unity/Aetherlight.Core/Presentation/Timeline.cs
using System;
using System.Collections.Generic;
using Aetherlight.Core;
using Aetherlight.Domain;

namespace Aetherlight.Presentation
{
    public enum NumeralStyle { Damage, Critical, Weak, Strong, Heal, Miss, Status, Aether }

    public enum BannerTone { Neutral, Good, Bad, Grand }

    public abstract class StepAction { }

    public sealed class PoseAction : StepAction { public string Pose = ""; }
    public sealed class LungeAction : StepAction { public string TowardId = ""; }
    public sealed class FlashAction : StepAction { public string Color = "#ffffff"; public double Intensity; }
    public sealed class ShakeAction : StepAction { public double Intensity; }
    public sealed class NumeralAction : StepAction { public string Text = ""; public NumeralStyle Style; }
    public sealed class BarAction : StepAction { public string Stat = "hp"; public double From; public double To; public double Max; }
    public sealed class HitstopAction : StepAction { }
    public sealed class BannerAction : StepAction { public string Text = ""; public BannerTone Tone; }
    public sealed class VfxAction : StepAction { public string Effect = ""; public Element? Element; }
    public sealed class CameraAction : StepAction { public string FocusId = ""; }
    public sealed class CollapseAction : StepAction { }
    public sealed class RiseAction : StepAction { }
    public sealed class ScreenFlashAction : StepAction { public string Color = "#ffffff"; }
    public sealed class SfxAction : StepAction { public string Cue = ""; }

    public sealed class AnimationStep
    {
        /// <summary>Entity id, or Timeline.GlobalTrack for scene-wide steps.</summary>
        public string Track = "";
        public double StartMs;
        public double DurationMs;
        public StepAction Action = new HitstopAction();
    }

    public sealed class Timeline
    {
        public List<AnimationStep> Steps = new List<AnimationStep>();
        public double DurationMs;
    }

    /// <summary>Beat lengths in milliseconds. Tune these to change the feel.</summary>
    public sealed class TimelineTiming
    {
        public double TurnStart = 160;
        public double WindUp = 240;
        public double Impact = 300;
        public double Hitstop = 90;
        public double NumeralFloat = 700;
        public double BetweenHits = 130;
        public double StatusApply = 320;
        public double MoteUnleash = 520;
        public double ClassChange = 600;
        public double SummonCutaway = 1400;
        public double Collapse = 520;
        public double Banner = 900;
        public double RoundGap = 120;

        public static TimelineTiming Default => new TimelineTiming();
    }

    /// <summary>Vitals at the start of the round, so bars animate from the right value.</summary>
    public sealed class VitalsSnapshot
    {
        public readonly Dictionary<string, (double Hp, double MaxHp, double Aether, double MaxAether)> Entries =
            new Dictionary<string, (double, double, double, double)>();

        public void Set(string id, double hp, double maxHp, double aether, double maxAether) =>
            Entries[id] = (hp, maxHp, aether, maxAether);
    }

    public sealed class TimelineOptions
    {
        public TimelineTiming Timing = TimelineTiming.Default;
        public VitalsSnapshot? Vitals;
        /// <summary>Display names for banners and numerals, keyed by id.</summary>
        public Dictionary<string, string> Labels = new Dictionary<string, string>();
    }

    /// <summary>
    /// Event stream to animation timeline.
    ///
    /// ResolveRound returns an entire round instantly as an ordered event list.
    /// This turns that list into a *timed* schedule: a flat array of steps, each
    /// with a start time, a duration and a track. Playback then becomes a matter
    /// of asking "which steps are active at time T", which any frame loop can
    /// drive.
    ///
    /// Deliberately pure - no clock, no engine types - so a schedule can be unit
    /// tested and ported without a renderer attached.
    ///
    /// Concurrency is expressed by giving steps overlapping time ranges. An
    /// all-foes spell hits every target in the same beat rather than
    /// sequentially, because the cursor only advances between beats.
    /// </summary>
    public static class TimelineBuilder
    {
        public const string GlobalTrack = "@global";

        public static Timeline Build(IReadOnlyList<GameEvent> events, TimelineOptions? options = null)
        {
            var opts = options ?? new TimelineOptions();
            var timing = opts.Timing;
            var timeline = new Timeline();

            // Track vitals as we walk so bar steps know their start value.
            var vitals = new Dictionary<string, (double Hp, double MaxHp, double Aether, double MaxAether)>();
            if (opts.Vitals != null)
            {
                foreach (var entry in opts.Vitals.Entries) vitals[entry.Key] = entry.Value;
            }

            double cursor = 0;

            void Emit(string track, double durationMs, StepAction action) =>
                timeline.Steps.Add(new AnimationStep { Track = track, StartMs = cursor, DurationMs = durationMs, Action = action });

            string Label(string id) => opts.Labels.TryGetValue(id, out var name) ? name : id;

            foreach (var e in events)
            {
                switch (e)
                {
                    case RoundStartEvent:
                        break;

                    case TurnStartEvent turn:
                        Emit(GlobalTrack, timing.TurnStart, new CameraAction { FocusId = turn.CombatantId });
                        cursor += timing.TurnStart;
                        break;

                    case TurnSkippedEvent skipped:
                        Emit(skipped.CombatantId, timing.NumeralFloat,
                            new NumeralAction { Text = ReadableReason(skipped.Reason), Style = NumeralStyle.Status });
                        cursor += timing.StatusApply;
                        break;

                    case ActionDeclaredEvent declared:
                    {
                        bool isAttack = declared.Action == "attack";
                        Emit(declared.CombatantId, timing.WindUp, new PoseAction { Pose = isAttack ? "attack" : "cast" });
                        if (isAttack && declared.TargetIds.Count > 0)
                            Emit(declared.CombatantId, timing.WindUp, new LungeAction { TowardId = declared.TargetIds[0] });
                        if (!isAttack && declared.Action.StartsWith("art:", StringComparison.Ordinal))
                            Emit(GlobalTrack, timing.Banner, new BannerAction { Text = Label(declared.Action.Substring(4)), Tone = BannerTone.Neutral });
                        cursor += timing.WindUp;
                        break;
                    }

                    case DamageEvent damage:
                    {
                        var style = DamageStyle(damage.Critical, damage.Effective);
                        Emit(damage.TargetId, timing.Impact, new FlashAction { Color = "#ffffff", Intensity = damage.Critical ? 1 : 0.75 });
                        Emit(damage.TargetId, timing.Impact, new ShakeAction { Intensity = damage.Critical ? 6 : 3 });
                        Emit(damage.TargetId, timing.Impact, new PoseAction { Pose = "hurt" });
                        Emit(damage.TargetId, timing.Impact, new VfxAction { Effect = "impact", Element = damage.Element });
                        Emit(damage.TargetId, timing.NumeralFloat, new NumeralAction { Text = FormatAmount(damage.Amount), Style = style });
                        Emit(GlobalTrack, timing.Hitstop, new HitstopAction());
                        Emit(GlobalTrack, 1, new SfxAction { Cue = damage.Critical ? "hit-critical" : "hit" });
                        EmitBar(timeline, vitals, damage.TargetId, "hp", -damage.Amount, cursor, timing.Impact);
                        cursor += timing.Impact + timing.BetweenHits;
                        break;
                    }

                    case HealEvent heal:
                        Emit(heal.TargetId, timing.Impact, new VfxAction { Effect = "heal" });
                        Emit(heal.TargetId, timing.NumeralFloat, new NumeralAction { Text = FormatAmount(heal.Amount), Style = NumeralStyle.Heal });
                        Emit(GlobalTrack, 1, new SfxAction { Cue = "heal" });
                        EmitBar(timeline, vitals, heal.TargetId, "hp", heal.Amount, cursor, timing.Impact);
                        cursor += timing.Impact;
                        break;

                    case MissEvent miss:
                        Emit(miss.TargetId, timing.NumeralFloat, new NumeralAction { Text = "MISS", Style = NumeralStyle.Miss });
                        Emit(GlobalTrack, 1, new SfxAction { Cue = "miss" });
                        cursor += timing.Impact;
                        break;

                    case AetherSpentEvent spent:
                        EmitBar(timeline, vitals, spent.CombatantId, "aether", -spent.Amount, cursor, timing.Impact);
                        break;

                    case AetherRestoredEvent restored:
                        Emit(restored.CombatantId, timing.NumeralFloat, new NumeralAction { Text = "+" + FormatAmount(restored.Amount), Style = NumeralStyle.Aether });
                        EmitBar(timeline, vitals, restored.CombatantId, "aether", restored.Amount, cursor, timing.Impact);
                        cursor += timing.Impact;
                        break;

                    case StatusAppliedEvent applied:
                        Emit(applied.TargetId, timing.StatusApply, new VfxAction { Effect = "status:" + applied.StatusId });
                        Emit(applied.TargetId, timing.NumeralFloat, new NumeralAction { Text = Label(applied.StatusId).ToUpperInvariant(), Style = NumeralStyle.Status });
                        cursor += timing.StatusApply;
                        break;

                    case StatusResistedEvent resisted:
                        Emit(resisted.TargetId, timing.NumeralFloat, new NumeralAction { Text = "RESIST", Style = NumeralStyle.Miss });
                        cursor += timing.StatusApply;
                        break;

                    case StatusExpiredEvent expired:
                        Emit(expired.TargetId, timing.StatusApply, new VfxAction { Effect = "status-clear" });
                        break;

                    case MoteUnleashedEvent unleashed:
                        Emit(unleashed.CombatantId, timing.MoteUnleash, new PoseAction { Pose = "cast" });
                        Emit(unleashed.CombatantId, timing.MoteUnleash, new VfxAction { Effect = "mote-unleash", Element = unleashed.Element });
                        Emit(GlobalTrack, timing.Banner, new BannerAction { Text = Label(unleashed.MoteId), Tone = BannerTone.Good });
                        Emit(GlobalTrack, 1, new SfxAction { Cue = "mote" });
                        cursor += timing.MoteUnleash;
                        break;

                    case MoteRecoveredEvent recovered:
                        Emit(recovered.CombatantId, timing.StatusApply, new VfxAction { Effect = "mote-return" });
                        break;

                    case ClassChangedEvent changed:
                        Emit(changed.CombatantId, timing.ClassChange, new VfxAction { Effect = "class-change" });
                        Emit(GlobalTrack, timing.Banner, new BannerAction { Text = Label(changed.ToClassId), Tone = BannerTone.Grand });
                        cursor += timing.ClassChange;
                        break;

                    case SummonCalledEvent summon:
                        Emit(GlobalTrack, timing.SummonCutaway, new BannerAction { Text = Label(summon.SummonId), Tone = BannerTone.Grand });
                        Emit(GlobalTrack, timing.SummonCutaway, new ScreenFlashAction { Color = "#ffffff" });
                        Emit(summon.CombatantId, timing.SummonCutaway, new PoseAction { Pose = "cast" });
                        Emit(GlobalTrack, 1, new SfxAction { Cue = "summon" });
                        cursor += timing.SummonCutaway;
                        break;

                    case DownedEvent downed:
                        Emit(downed.CombatantId, timing.Collapse, new PoseAction { Pose = "down" });
                        Emit(downed.CombatantId, timing.Collapse, new CollapseAction());
                        Emit(GlobalTrack, 1, new SfxAction { Cue = "down" });
                        cursor += timing.Collapse;
                        break;

                    case RevivedEvent revived:
                        Emit(revived.CombatantId, timing.Collapse, new RiseAction());
                        Emit(revived.CombatantId, timing.NumeralFloat, new NumeralAction { Text = FormatAmount(revived.Hp), Style = NumeralStyle.Heal });
                        SetVital(vitals, revived.CombatantId, "hp", revived.Hp);
                        cursor += timing.Collapse;
                        break;

                    case FleeSucceededEvent:
                        Emit(GlobalTrack, timing.Banner, new BannerAction { Text = "Escaped", Tone = BannerTone.Neutral });
                        cursor += timing.Banner;
                        break;

                    case FleeFailedEvent:
                        Emit(GlobalTrack, timing.Banner, new BannerAction { Text = "Couldn't escape", Tone = BannerTone.Bad });
                        cursor += timing.Banner;
                        break;

                    case BattleEndedEvent ended:
                    {
                        var tone = ended.Outcome switch
                        {
                            BattleOutcome.Victory => BannerTone.Good,
                            BattleOutcome.Defeat => BannerTone.Bad,
                            _ => BannerTone.Neutral,
                        };
                        Emit(GlobalTrack, timing.Banner, new BannerAction { Text = OutcomeText(ended.Outcome), Tone = tone });
                        cursor += timing.Banner;
                        break;
                    }

                    case RewardEvent reward:
                        Emit(GlobalTrack, timing.Banner, new BannerAction { Text = $"+{reward.Xp} XP   +{reward.Coin}", Tone = BannerTone.Good });
                        cursor += timing.Banner;
                        break;

                    case MessageEvent message:
                        Emit(GlobalTrack, timing.Banner, new BannerAction { Text = message.Text, Tone = BannerTone.Neutral });
                        cursor += timing.Banner;
                        break;

                    case RoundEndEvent:
                        cursor += timing.RoundGap;
                        break;
                }
            }

            double duration = 0;
            foreach (var step in timeline.Steps) duration = Math.Max(duration, step.StartMs + step.DurationMs);
            timeline.DurationMs = duration;
            return timeline;
        }

        private static void EmitBar(
            Timeline timeline,
            Dictionary<string, (double Hp, double MaxHp, double Aether, double MaxAether)> vitals,
            string id,
            string stat,
            double delta,
            double startMs,
            double durationMs)
        {
            if (!vitals.TryGetValue(id, out var entry)) return;

            double max = stat == "hp" ? entry.MaxHp : entry.MaxAether;
            double from = stat == "hp" ? entry.Hp : entry.Aether;
            double to = Math.Max(0, Math.Min(max, from + delta));

            vitals[id] = stat == "hp"
                ? (to, entry.MaxHp, entry.Aether, entry.MaxAether)
                : (entry.Hp, entry.MaxHp, to, entry.MaxAether);

            timeline.Steps.Add(new AnimationStep
            {
                Track = id,
                StartMs = startMs,
                DurationMs = durationMs,
                Action = new BarAction { Stat = stat, From = from, To = to, Max = max },
            });
        }

        private static void SetVital(
            Dictionary<string, (double Hp, double MaxHp, double Aether, double MaxAether)> vitals,
            string id,
            string stat,
            double value)
        {
            if (!vitals.TryGetValue(id, out var entry)) return;
            vitals[id] = stat == "hp"
                ? (value, entry.MaxHp, entry.Aether, entry.MaxAether)
                : (entry.Hp, entry.MaxHp, value, entry.MaxAether);
        }

        /// <summary>
        /// Numeral styling from the damage event's own data.
        ///
        /// Effective is the elemental multiplier the rules layer already worked
        /// out, which is exactly why it is on the event: the presentation layer
        /// should not be recomputing affinity to decide how a hit reads.
        /// </summary>
        public static NumeralStyle DamageStyle(bool critical, double effective)
        {
            if (critical) return NumeralStyle.Critical;
            if (effective >= 1.2) return NumeralStyle.Strong;
            if (effective <= 0.8) return NumeralStyle.Weak;
            return NumeralStyle.Damage;
        }

        /// <summary>Amounts are whole numbers on screen even though they are doubles.</summary>
        private static string FormatAmount(double amount) =>
            ((long)Math.Round(amount, MidpointRounding.AwayFromZero)).ToString(System.Globalization.CultureInfo.InvariantCulture);

        private static string OutcomeText(BattleOutcome outcome) => outcome switch
        {
            BattleOutcome.Victory => "Victory",
            BattleOutcome.Defeat => "Defeat",
            _ => "Escaped",
        };

        private static string ReadableReason(string reason) => reason switch
        {
            "insufficient-aether" => "NO AETHER",
            "insufficient-motes" => "NO MOTES",
            "arts-sealed" => "SEALED",
            _ => reason.Replace("-", " ").ToUpperInvariant(),
        };
    }
}

File: src/presentation/timeline.ts
/**
 * Event stream to animation timeline.
 *
 * `resolveRound()` returns an entire round instantly as an ordered event list.
 * This module turns that list into a *timed* schedule: a flat array of steps,
 * each with a start time, a duration, and a track. Playback then becomes a
 * matter of asking "which steps are active at time T", which is trivial to
 * drive from any engine's frame loop.
 *
 * This is the load-bearing piece of the presentation contract, and it is
 * deliberately pure - no canvas, no engine types, no clock. Given the same
 * events it produces the same schedule every time, which means it can be unit
 * tested and ported without a renderer attached.
 *
 * Concurrency is expressed by giving steps overlapping time ranges. An
 * all-foes spell hits every target in the same beat rather than sequentially,
 * because the cursor only advances between beats.
 */

import type { GameEvent } from '../core/events.js';
import type { Element } from '../domain/elements.js';

export const GLOBAL_TRACK = '@global';

export type StepAction =
  /** Play a named pose on an actor. */
  | { kind: 'pose'; pose: string }
  /** Lunge toward a target and return. */
  | { kind: 'lunge'; towardId: string }
  /** Tint flash, used for impacts. */
  | { kind: 'flash'; color: string; intensity: number }
  /** Positional shake. */
  | { kind: 'shake'; intensity: number }
  /** A floating number or word. */
  | { kind: 'numeral'; text: string; style: NumeralStyle }
  /** Animate a stat bar between two values. */
  | { kind: 'bar'; stat: 'hp' | 'aether'; from: number; to: number; max: number }
  /** Freeze the whole scene briefly to sell an impact. */
  | { kind: 'hitstop' }
  /** Full-width text banner. */
  | { kind: 'banner'; text: string; tone: 'neutral' | 'good' | 'bad' | 'grand' }
  /** An elemental effect burst at an entity. */
  | { kind: 'vfx'; effect: string; element: Element | null }
  /** Point the camera at an entity. */
  | { kind: 'camera'; focusId: string }
  /** Fade an entity out as it falls. */
  | { kind: 'collapse' }
  /** Fade an entity back in. */
  | { kind: 'rise' }
  /** Screen-wide flash, for summons. */
  | { kind: 'screen-flash'; color: string }
  /** Named sound cue; the host decides what it maps to. */
  | { kind: 'sfx'; cue: string };

export type NumeralStyle = 'damage' | 'critical' | 'weak' | 'strong' | 'heal' | 'miss' | 'status' | 'aether';

export interface AnimationStep {
  /** Entity id, or GLOBAL_TRACK for scene-wide steps. */
  track: string;
  startMs: number;
  durationMs: number;
  action: StepAction;
}

export interface Timeline {
  steps: AnimationStep[];
  durationMs: number;
}

/** Beat lengths, in milliseconds. Tune these to change the whole feel. */
export interface TimelineTiming {
  turnStart: number;
  windUp: number;
  impact: number;
  hitstop: number;
  numeralFloat: number;
  betweenHits: number;
  statusApply: number;
  moteUnleash: number;
  classChange: number;
  summonCutaway: number;
  collapse: number;
  banner: number;
  roundGap: number;
}

export const DEFAULT_TIMING: TimelineTiming = Object.freeze({
  turnStart: 160,
  windUp: 240,
  impact: 300,
  hitstop: 90,
  numeralFloat: 700,
  betweenHits: 130,
  statusApply: 320,
  moteUnleash: 520,
  classChange: 600,
  summonCutaway: 1400,
  collapse: 520,
  banner: 900,
  roundGap: 120,
});

/** Vitals at the start of the round, so bars can animate from the right value. */
export interface VitalsSnapshot {
  [combatantId: string]: { hp: number; maxHp: number; aether: number; maxAether: number };
}

export interface TimelineOptions {
  timing?: Partial<TimelineTiming>;
  vitals?: VitalsSnapshot;
  /** Names for banners and numerals, keyed by id. */
  labels?: Record<string, string>;
}

/**
 * Build a timeline from one round's events.
 *
 * The cursor advances beat by beat. Steps that should be simultaneous are
 * emitted at the same cursor value; the cursor only moves when the next thing
 * genuinely follows the last.
 */
export function buildTimeline(events: readonly GameEvent[], options: TimelineOptions = {}): Timeline {
  const timing: TimelineTiming = { ...DEFAULT_TIMING, ...options.timing };
  const labels = options.labels ?? {};
  const steps: AnimationStep[] = [];

  // Track vitals as we walk so bar steps know their start value.
  const vitals: VitalsSnapshot = {};
  for (const [id, value] of Object.entries(options.vitals ?? {})) vitals[id] = { ...value };

  let cursor = 0;
  const emit = (step: AnimationStep): void => {
    steps.push(step);
  };
  const label = (id: string): string => labels[id] ?? id;

  for (const event of events) {
    switch (event.type) {
      case 'round-start':
        break;

      case 'turn-start': {
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.turnStart, action: { kind: 'camera', focusId: event.combatantId } });
        cursor += timing.turnStart;
        break;
      }

      case 'turn-skipped': {
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.numeralFloat, action: { kind: 'numeral', text: readableReason(event.reason), style: 'status' } });
        cursor += timing.statusApply;
        break;
      }

      case 'action-declared': {
        const isAttack = event.action === 'attack';
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.windUp, action: { kind: 'pose', pose: isAttack ? 'attack' : 'cast' } });
        if (isAttack && event.targetIds[0]) {
          emit({ track: event.combatantId, startMs: cursor, durationMs: timing.windUp, action: { kind: 'lunge', towardId: event.targetIds[0] } });
        }
        if (!isAttack && event.action.startsWith('art:')) {
          emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.banner, action: { kind: 'banner', text: label(event.action.slice(4)), tone: 'neutral' } });
        }
        cursor += timing.windUp;
        break;
      }

      case 'damage': {
        const style = damageStyle(event.critical, event.effective);
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.impact, action: { kind: 'flash', color: '#ffffff', intensity: event.critical ? 1 : 0.75 } });
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.impact, action: { kind: 'shake', intensity: event.critical ? 6 : 3 } });
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.impact, action: { kind: 'pose', pose: 'hurt' } });
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.impact, action: { kind: 'vfx', effect: 'impact', element: event.element } });
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.numeralFloat, action: { kind: 'numeral', text: String(event.amount), style } });
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.hitstop, action: { kind: 'hitstop' } });
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: 1, action: { kind: 'sfx', cue: event.critical ? 'hit-critical' : 'hit' } });
        emitBar(emit, vitals, event.targetId, 'hp', -event.amount, cursor, timing.impact);
        cursor += timing.impact + timing.betweenHits;
        break;
      }

      case 'heal': {
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.impact, action: { kind: 'vfx', effect: 'heal', element: null } });
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.numeralFloat, action: { kind: 'numeral', text: String(event.amount), style: 'heal' } });
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: 1, action: { kind: 'sfx', cue: 'heal' } });
        emitBar(emit, vitals, event.targetId, 'hp', event.amount, cursor, timing.impact);
        cursor += timing.impact;
        break;
      }

      case 'miss': {
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.numeralFloat, action: { kind: 'numeral', text: 'MISS', style: 'miss' } });
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: 1, action: { kind: 'sfx', cue: 'miss' } });
        cursor += timing.impact;
        break;
      }

      case 'aether-spent':
        emitBar(emit, vitals, event.combatantId, 'aether', -event.amount, cursor, timing.impact);
        break;

      case 'aether-restored': {
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.numeralFloat, action: { kind: 'numeral', text: `+${event.amount}`, style: 'aether' } });
        emitBar(emit, vitals, event.combatantId, 'aether', event.amount, cursor, timing.impact);
        cursor += timing.impact;
        break;
      }

      case 'status-applied': {
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.statusApply, action: { kind: 'vfx', effect: `status:${event.statusId}`, element: null } });
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.numeralFloat, action: { kind: 'numeral', text: label(event.statusId).toUpperCase(), style: 'status' } });
        cursor += timing.statusApply;
        break;
      }

      case 'status-resisted': {
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.numeralFloat, action: { kind: 'numeral', text: 'RESIST', style: 'miss' } });
        cursor += timing.statusApply;
        break;
      }

      case 'status-expired':
        emit({ track: event.targetId, startMs: cursor, durationMs: timing.statusApply, action: { kind: 'vfx', effect: 'status-clear', element: null } });
        break;

      case 'mote-unleashed': {
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.moteUnleash, action: { kind: 'pose', pose: 'cast' } });
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.moteUnleash, action: { kind: 'vfx', effect: 'mote-unleash', element: event.element } });
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.banner, action: { kind: 'banner', text: label(event.moteId), tone: 'good' } });
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: 1, action: { kind: 'sfx', cue: 'mote' } });
        cursor += timing.moteUnleash;
        break;
      }

      case 'mote-recovered':
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.statusApply, action: { kind: 'vfx', effect: 'mote-return', element: null } });
        break;

      case 'class-changed': {
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.classChange, action: { kind: 'vfx', effect: 'class-change', element: null } });
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.banner, action: { kind: 'banner', text: label(event.toClassId), tone: 'grand' } });
        cursor += timing.classChange;
        break;
      }

      case 'summon-called': {
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.summonCutaway, action: { kind: 'banner', text: label(event.summonId), tone: 'grand' } });
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.summonCutaway, action: { kind: 'screen-flash', color: '#ffffff' } });
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.summonCutaway, action: { kind: 'pose', pose: 'cast' } });
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: 1, action: { kind: 'sfx', cue: 'summon' } });
        cursor += timing.summonCutaway;
        break;
      }

      case 'downed': {
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.collapse, action: { kind: 'pose', pose: 'down' } });
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.collapse, action: { kind: 'collapse' } });
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: 1, action: { kind: 'sfx', cue: 'down' } });
        cursor += timing.collapse;
        break;
      }

      case 'revived': {
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.collapse, action: { kind: 'rise' } });
        emit({ track: event.combatantId, startMs: cursor, durationMs: timing.numeralFloat, action: { kind: 'numeral', text: String(event.hp), style: 'heal' } });
        setVital(vitals, event.combatantId, 'hp', event.hp);
        cursor += timing.collapse;
        break;
      }

      case 'flee-succeeded':
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.banner, action: { kind: 'banner', text: 'Escaped', tone: 'neutral' } });
        cursor += timing.banner;
        break;

      case 'flee-failed':
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.banner, action: { kind: 'banner', text: "Couldn't escape", tone: 'bad' } });
        cursor += timing.banner;
        break;

      case 'battle-ended': {
        const tone = event.outcome === 'victory' ? 'good' : event.outcome === 'defeat' ? 'bad' : 'neutral';
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.banner, action: { kind: 'banner', text: outcomeText(event.outcome), tone } });
        cursor += timing.banner;
        break;
      }

      case 'reward':
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.banner, action: { kind: 'banner', text: `+${event.xp} XP   +${event.coin}`, tone: 'good' } });
        cursor += timing.banner;
        break;

      case 'message':
        emit({ track: GLOBAL_TRACK, startMs: cursor, durationMs: timing.banner, action: { kind: 'banner', text: event.text, tone: 'neutral' } });
        cursor += timing.banner;
        break;

      case 'round-end':
        cursor += timing.roundGap;
        break;

      default:
        break;
    }
  }

  const durationMs = steps.reduce((max, step) => Math.max(max, step.startMs + step.durationMs), 0);
  return { steps, durationMs };
}

function emitBar(
  emit: (step: AnimationStep) => void,
  vitals: VitalsSnapshot,
  id: string,
  stat: 'hp' | 'aether',
  delta: number,
  startMs: number,
  durationMs: number,
): void {
  const entry = vitals[id];
  if (!entry) return;
  const max = stat === 'hp' ? entry.maxHp : entry.maxAether;
  const from = stat === 'hp' ? entry.hp : entry.aether;
  const to = Math.max(0, Math.min(max, from + delta));
  if (stat === 'hp') entry.hp = to;
  else entry.aether = to;
  emit({ track: id, startMs, durationMs, action: { kind: 'bar', stat, from, to, max } });
}

function setVital(vitals: VitalsSnapshot, id: string, stat: 'hp' | 'aether', value: number): void {
  const entry = vitals[id];
  if (!entry) return;
  if (stat === 'hp') entry.hp = value;
  else entry.aether = value;
}

/**
 * Numeral styling from the damage event's own data.
 *
 * `effective` is the elemental multiplier the rules layer already computed,
 * which is exactly why it is on the event: the presentation layer should not
 * be recomputing affinity to decide how a hit reads.
 */
export function damageStyle(critical: boolean, effective: number): NumeralStyle {
  if (critical) return 'critical';
  if (effective >= 1.2) return 'strong';
  if (effective <= 0.8) return 'weak';
  return 'damage';
}

function outcomeText(outcome: 'victory' | 'defeat' | 'fled'): string {
  if (outcome === 'victory') return 'Victory';
  if (outcome === 'defeat') return 'Defeat';
  return 'Escaped';
}

function readableReason(reason: string): string {
  switch (reason) {
    case 'insufficient-aether':
      return 'NO AETHER';
    case 'insufficient-motes':
      return 'NO MOTES';
    case 'arts-sealed':
      return 'SEALED';
    default:
      return reason.replace(/-/g, ' ').toUpperCase();
  }
}
